Konamiman/Nextor
# Improve the performance of cluster chain allocation

On FAT16 drives that have 64K clusters and are almost full, allocating new clusters is a very slow process. The entire FAT is scanned every time, and copying big files, a process that involves many allocations, takes an absurd amount of time. As an user mentioned in [the msx.org forum](https://msx.org/forum/msx-talk/development/my-plans-for-nextor-v3-warning-broken-drivers-ahead):

> I once tested with 4Mb partition filled 75% and copying 16Mb file into it - after one hour I just turned computer off - it was still copying stuff

This pull request introduces a mechanism that dramatically improves the performance of the cluster chain allocation process: a per-drive cache of the last cluster number allocated that prevents the search process from starting over at the beginning of the FAT every time.

## How is it implemented

A new field named `UD_ACLU` is added to the unit descriptor (a data structure that exists for each mapped drive). This field stores the cluster number where the search for free clusters will start the next time a cluster chain needs to be allocated (minus one). It's initialized to 1 when the descriptor is created (at boot time, or when the drive is explicitly mapped at a later time) and modified as follows:

- After a cluster chain allocation process succeeds, it's set to the number of the last cluster that has been allocated.

- When a cluster chain is freed, it's set to the number of the lowest cluster number from the chain that has been freed, minus one, but **only** if that new value would be lower than the previous value.

This algorithm guarantees that the allocation process works as usual (the lowest numbered free clusters are allocated first), but much faster.

Additionally, `UD_ACLU` is initialized to 1 again if:

- The allocation of a cluster chain fails with a "Disk full" error (or with any other error, actually, but "Disk full" will be the most common case). This is a fail safe mechanism: if the algorithm has any flaw and trying to write to the drive causes a "Disk full" error even though there is actually some free space available, the next attempt will succeed.

- The drive is written to by using the raw sector write `RDABS` or `WRDRV` function calls. This is to account for the case where a disk optimizer tool writes to the FAT sectors directly; the number of the sectors being written is not checked to keep the code simple (and direct sector writing using these functions is an uncommon case after all).

## How to test

The easiest way to test is to use an emulator.

1. Create a disk image with a size that is a power of 2, so that it gets the maximum number of clusters (close to 64K). A size of 128MB will be enough.

2. Fill the disk image with files until it's almost completely full (leave e.g. about 100K free).

3. Boot with that disk image, go to BASIC and create a trivial program e.g. `10 REM`. Save it, `save"x"`. It will take about 5s for a R800 or about 30s for a Z80.

4. Save the program again with a different name, notice how the process is now immediate.

If you want to try the "copy a 16MB file" scenario instead, you'll see that now a R800 needs about two minutes to complete the process, which is about the same time as when the drive is empty.
